sdk/scheduler/src/test/java/com/mesosphere/sdk/scheduler/TokenBucketTest.java

```diff
@@ -0,0 +1,120 @@
+package com.mesosphere.sdk.scheduler;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.time.Duration;
+
+/**
+ * This class tests the {@link TokenBucket}.
+ */
+public class TokenBucketTest {
+    @Test
+    public void acquireFailsOnEmptyBucket() {
+        TokenBucket bucket = TokenBucket.newBuilder()
+                .initialTokenCount(0)
+                .build();
+        Assert.assertFalse(bucket.tryAcquire());
+    }
+
+    @Test
+    public void acquireSucceedsOnSingleTokenBucker() {
+        TokenBucket bucket = TokenBucket.newBuilder()
+                .initialTokenCount(1)
+                .build();
+        Assert.assertTrue(bucket.tryAcquire());
+    }
+
+    @Test
+    public void acquireFailsOnTooFastAcquire() {
+        TokenBucket bucket = TokenBucket.newBuilder()
+                .initialTokenCount(2)
+                .build();
+        Assert.assertTrue(bucket.tryAcquire());
+        Assert.assertFalse(bucket.tryAcquire());
+    }
+
+    @Test
+    public void allowSuperFastAcquire() {
+        TokenBucket bucket = TokenBucket.newBuilder()
+                .initialTokenCount(2)
+                .acquireInterval(Duration.ZERO)
+                .build();
+        Assert.assertTrue(bucket.tryAcquire());
+        Assert.assertTrue(bucket.tryAcquire());
+    }
+
+    @Test
+    public void exhaustTokens() {
+        TokenBucket bucket = TokenBucket.newBuilder()
+                .initialTokenCount(1)
+                .acquireInterval(Duration.ZERO)
+                .build();
+        Assert.assertTrue(bucket.tryAcquire());
+        Assert.assertFalse(bucket.tryAcquire());
+    }
+
+    @Test
+    public void replenishTokens() throws InterruptedException {
+        Duration incrementInterval = Duration.ofMillis(100);
+        TokenBucket bucket = TokenBucket.newBuilder()
+                .initialTokenCount(1)
+                .acquireInterval(Duration.ZERO)
+                .incrementInterval(incrementInterval)
+                .build();
+
+        Assert.assertTrue(bucket.tryAcquire());
+        Assert.assertFalse(bucket.tryAcquire());
+        Thread.sleep(incrementInterval.toMillis() + 50); // Give 50ms of breathing room to the token incrementing thread
```

Comment:
Passage of time should be simulated by doing the following:
- Adding a `protected now()` function to `TokenBucket` which returns `System.currentTimeMillis()`
- Updating all internal calls to `System.currentTimeMillis()` in `TokenBucket` with calls to `now()`
- In these tests, have a `TestTokenBucket` implementation which overrides `now()`, and has a function to allow incrementing the value to be returned by `now()`.